File: packages/edit-plan-schema/python/edit_plan_schema/schema.py
"""The edit-plan JSON contract (ADR-0005). Claude's output, the render
worker's input, and the frontend timeline all derive from this shape.

`timeline` is the sole rendering authority. Every other timed array uses
output-timeline coordinates (post-cut), not source-video coordinates. See
docs/adr/0005-edit-plan-schema.md for the full reasoning.

This model validates *raw* Claude output, before validate_and_clamp() has
run. Timestamp fields (source_start, output_end, start, end, at, ...)
deliberately carry no ge=0/duration bound here -- that range depends on the
actual source video's duration, which this layer doesn't know, and
hallucinated out-of-range timestamps are exactly what validate_and_clamp()
exists to repair rather than reject. Only clamp-proof invariants (types,
scores, enums, zoom scale) are enforced at this layer.
"""

import uuid
from typing import Literal

from pydantic import BaseModel, Field

CutReason = Literal["silence", "filler_word", "dead_moment", "repeated_section", "other"]
TransitionType = Literal["hard_cut", "crossfade", "whip_pan", "zoom_punch"]
ZoomFocus = Literal["center", "left", "right", "top", "bottom"]
MotionGraphicTemplate = Literal[
    "lower_third", "callout", "counter", "cta_screen", "highlight_box", "arrow", "progress_bar"
]
BrollSource = Literal["pexels", "pixabay"]


class SourceVideoRef(BaseModel):
    source_video_id: uuid.UUID
    duration_seconds: float = Field(ge=0)


class Clip(BaseModel):
    """One entry in `timeline`. Rendering authority (ADR-0005) -- ffmpeg's
    cut/concat step is driven entirely by this array, in order."""

    id: str
    source_video_id: uuid.UUID
    source_start: float
    source_end: float
    output_start: float
    output_end: float


class CutRange(BaseModel):
    """Informational only (ADR-0005) -- a range removed from a source video,
    for UI display ("removed 45s of silence") and audit, never read by the
    renderer."""

    source_video_id: uuid.UUID
    source_start: float
    source_end: float
    reason: CutReason = "other"


class Subtitle(BaseModel):
    start: float
    end: float
    text: str
    emphasis_words: list[str] = Field(default_factory=list)


class Zoom(BaseModel):
    start: float
    end: float
    scale: float = Field(gt=0, le=3.0)
    focus: ZoomFocus = "center"


class Transition(BaseModel):
    at: float
    type: TransitionType
    duration_seconds: float = Field(default=0.0, ge=0)


class SoundEffect(BaseModel):
    at: float
    asset_tag: str
    volume_db: float = 0.0


class MotionGraphic(BaseModel):
    start: float
    end: float
    template: MotionGraphicTemplate
    props: dict = Field(default_factory=dict)


class BrollClip(BaseModel):
    start: float
    end: float
    keyword: str
    source: BrollSource = "pexels"


class MusicCue(BaseModel):
    start: float
    end: float
    asset_tag: str
    volume_db: float = -18.0


class ViralScore(BaseModel):
    hook_score: int = Field(ge=0, le=100)
    retention_score: int = Field(ge=0, le=100)
    engagement_score: int = Field(ge=0, le=100)
    reasoning: str = ""


class RetentionPrediction(BaseModel):
    predicted_drop_off_points: list[float] = Field(default_factory=list)
    notes: str = ""


class EditPlan(BaseModel):
    version: Literal["1.0"] = "1.0"
    style_slug: str
    source_videos: list[SourceVideoRef]

    timeline: list[Clip]
    cuts: list[CutRange] = Field(default_factory=list)
    subtitles: list[Subtitle] = Field(default_factory=list)
    zooms: list[Zoom] = Field(default_factory=list)
    transitions: list[Transition] = Field(default_factory=list)
    sound_effects: list[SoundEffect] = Field(default_factory=list)
    motion_graphics: list[MotionGraphic] = Field(default_factory=list)
    broll: list[BrollClip] = Field(default_factory=list)
    music: list[MusicCue] = Field(default_factory=list)

    viral_score: ViralScore
    retention_prediction: RetentionPrediction = Field(default_factory=RetentionPrediction)

File: packages/edit-plan-schema/node/src/schema.ts
/**
 * The edit-plan JSON contract (ADR-0005). Mirrors
 * packages/edit-plan-schema/python/edit_plan_schema/schema.py field for
 * field -- keep both in sync by hand when either changes.
 *
 * `timeline` is the sole rendering authority. Every other timed array uses
 * output-timeline coordinates (post-cut), not source-video coordinates.
 *
 * This is a structural-validation guard only. Clamping (out-of-range
 * timestamps, re-deriving output_start/output_end) happens on the Python
 * side before a plan ever reaches this schema -- see
 * docs/adr/0005-edit-plan-schema.md.
 */

import { z } from "zod";

const nonNegative = z.number().min(0);

export const CutReasonSchema = z.enum([
  "silence",
  "filler_word",
  "dead_moment",
  "repeated_section",
  "other",
]);

export const TransitionTypeSchema = z.enum([
  "hard_cut",
  "crossfade",
  "whip_pan",
  "zoom_punch",
]);

export const ZoomFocusSchema = z.enum(["center", "left", "right", "top", "bottom"]);

export const MotionGraphicTemplateSchema = z.enum([
  "lower_third",
  "callout",
  "counter",
  "cta_screen",
  "highlight_box",
  "arrow",
  "progress_bar",
]);

export const BrollSourceSchema = z.enum(["pexels", "pixabay"]);

export const SourceVideoRefSchema = z.object({
  source_video_id: z.string().uuid(),
  duration_seconds: nonNegative,
});

export const ClipSchema = z.object({
  id: z.string(),
  source_video_id: z.string().uuid(),
  source_start: nonNegative,
  source_end: nonNegative,
  output_start: nonNegative,
  output_end: nonNegative,
});

export const CutRangeSchema = z.object({
  source_video_id: z.string().uuid(),
  source_start: nonNegative,
  source_end: nonNegative,
  reason: CutReasonSchema.default("other"),
});

export const SubtitleSchema = z.object({
  start: nonNegative,
  end: nonNegative,
  text: z.string(),
  emphasis_words: z.array(z.string()).default([]),
});

export const ZoomSchema = z.object({
  start: nonNegative,
  end: nonNegative,
  scale: z.number().gt(0).lte(3.0),
  focus: ZoomFocusSchema.default("center"),
});

export const TransitionSchema = z.object({
  at: nonNegative,
  type: TransitionTypeSchema,
  duration_seconds: nonNegative.default(0),
});

export const SoundEffectSchema = z.object({
  at: nonNegative,
  asset_tag: z.string(),
  volume_db: z.number().default(0),
});

export const MotionGraphicSchema = z.object({
  start: nonNegative,
  end: nonNegative,
  template: MotionGraphicTemplateSchema,
  props: z.record(z.string(), z.unknown()).default({}),
});

export const BrollClipSchema = z.object({
  start: nonNegative,
  end: nonNegative,
  keyword: z.string(),
  source: BrollSourceSchema.default("pexels"),
});

export const MusicCueSchema = z.object({
  start: nonNegative,
  end: nonNegative,
  asset_tag: z.string(),
  volume_db: z.number().default(-18.0),
});

export const ViralScoreSchema = z.object({
  hook_score: z.number().int().min(0).max(100),
  retention_score: z.number().int().min(0).max(100),
  engagement_score: z.number().int().min(0).max(100),
  reasoning: z.string().default(""),
});

export const RetentionPredictionSchema = z.object({
  predicted_drop_off_points: z.array(z.number()).default([]),
  notes: z.string().default(""),
});

export const EditPlanSchema = z.object({
  version: z.literal("1.0").default("1.0"),
  style_slug: z.string(),
  source_videos: z.array(SourceVideoRefSchema),

  timeline: z.array(ClipSchema).min(1),
  cuts: z.array(CutRangeSchema).default([]),
  subtitles: z.array(SubtitleSchema).default([]),
  zooms: z.array(ZoomSchema).default([]),
  transitions: z.array(TransitionSchema).default([]),
  sound_effects: z.array(SoundEffectSchema).default([]),
  motion_graphics: z.array(MotionGraphicSchema).default([]),
  broll: z.array(BrollClipSchema).default([]),
  music: z.array(MusicCueSchema).default([]),

  viral_score: ViralScoreSchema,
  retention_prediction: RetentionPredictionSchema.default({
    predicted_drop_off_points: [],
    notes: "",
  }),
});

export type EditPlan = z.infer<typeof EditPlanSchema>;
export type Clip = z.infer<typeof ClipSchema>;
export type Subtitle = z.infer<typeof SubtitleSchema>;
export type Zoom = z.infer<typeof ZoomSchema>;
export type Transition = z.infer<typeof TransitionSchema>;
export type SoundEffect = z.infer<typeof SoundEffectSchema>;
export type MotionGraphic = z.infer<typeof MotionGraphicSchema>;
export type BrollClip = z.infer<typeof BrollClipSchema>;
export type MusicCue = z.infer<typeof MusicCueSchema>;
export type ViralScore = z.infer<typeof ViralScoreSchema>;
export type RetentionPrediction = z.infer<typeof RetentionPredictionSchema>;
